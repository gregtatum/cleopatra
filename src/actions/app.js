/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import {
  getAreTooltipsEnabled,
  getIsTooltipDismissRequested,
} from '../reducers/app';
import { getSelectedTab, getDataSource } from '../reducers/url-state';
import { sendAnalytics } from '../utils/analytics';
import type { Action, ThunkAction } from '../types/store';
import type { TabSlug } from '../app-logic/tabs-handling';
import type { ProfileSharingStatus, UrlState } from '../types/reducers';
import type { TooltipReference, MousePosition } from '../types/actions';
import type { CssPixels } from '../types/units';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function setProfileSharingStatus(
  profileSharingStatus: ProfileSharingStatus
): Action {
  return {
    type: 'SET_PROFILE_SHARING_STATUS',
    profileSharingStatus,
  };
}

export function changeTabOrder(tabOrder: number[]): Action {
  return {
    type: 'CHANGE_TAB_ORDER',
    tabOrder,
  };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

/**
 * Tooltips are created from global Redux state. This action causes one to show
 * up in the UI. The action is ignored if tooltips are currently disabled.
 */
export function viewTooltip(
  mouseX: CssPixels,
  mouseY: CssPixels,
  tooltipReference: TooltipReference
): ThunkAction<void> {
  return (dispatch, getState) => {
    if (getAreTooltipsEnabled(getState())) {
      dispatch({
        type: 'VIEW_TOOLTIP',
        tooltipReference,
        mouse: { mouseX, mouseY },
      });
    }
  };
}

/**
 * Dismiss the tooltip immediately.
 */
export function dismissTooltip(): Action {
  return {
    type: 'DISMISS_TOOLTIP',
  };
}

/**
 * Re-position the tooltip.
 */
export function moveTooltip(mouse: MousePosition): Action {
  return {
    type: 'MOVE_TOOLTIP',
    mouse,
  };
}

/**
 * Temporarily disable tooltips from showing. This is useful for stopping tooltips
 * from interfering with mouse actions like draggin.
 */
export function disableTooltips(): Action {
  return {
    type: 'DISABLE_TOOLTIPS',
  };
}

/**
 * Re-enable a tooltip after disabling it.
 */
export function enableTooltips(): Action {
  return {
    type: 'ENABLE_TOOLTIPS',
  };
}

/**
 * It is difficult to coordinate the closing of tooltips between separate components.
 * One may fire the mouseleave, and one will fire the mouseenter. Defer the closing
 * of the tooltip, and initially only request it to be dismissed.
 */
export function requestToDismissTooltip(): ThunkAction<void> {
  return (dispatch, getState) => {
    const { generation: oldGeneration } = getIsTooltipDismissRequested(
      getState()
    );
    const generation = oldGeneration + 1;
    dispatch({
      type: 'REQUEST_TO_DISMISS_TOOLTIP',
      generation,
    });

    // Wait until the next animation frame to allow all of the current event listeners
    // to fire and decide what they want to do.
    requestAnimationFrame(() => {
      const {
        isRequested,
        generation: currentGeneration,
      } = getIsTooltipDismissRequested(getState());

      // Check that the request to
      if (isRequested && generation === currentGeneration) {
        dispatch(dismissTooltip());
      }
    });
  };
}

/**
 * It is difficult to coordinate the closing of tooltips between separate components.
 * One may fire the mouseleave, and one will fire the mouseenter. This action stops
 * the request to dismiss a tooltip, and leaves it open.
 */
export function keepTooltipOpen(): Action {
  return {
    type: 'KEEP_TOOLTIP_OPEN',
  };
}
