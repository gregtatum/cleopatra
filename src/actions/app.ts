/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { getSelectedTab, getDataSource } from '../selectors/url-state';
import { getTrackThreadHeights } from '../selectors/app';
import { sendAnalytics } from '../utils/analytics';
import { stateFromLocation } from '../app-logic/url-handling';
import { finalizeProfileView } from './receive-profile';
import { Profile, ThreadIndex } from '../types/profile';
import { CssPixels } from '../types/units';
import { Action, ThunkAction } from '../types/store';
import { TabSlug } from '../app-logic/tabs-handling';
import { UrlState } from '../types/state';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function startFetchingProfiles(): Action {
  return { type: 'START_FETCHING_PROFILES' };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when we start setting up the initial url state.
 * It takes the location and profile data, converts the location into url
 * state and then dispatches relevant actions to finalize the view.
 */
export function setupInitialUrlState(
  location: Location,
  profile: Profile
): ThunkAction<void> {
  return dispatch => {
    let urlState;
    try {
      urlState = stateFromLocation(location, profile);
    } catch (e) {
      // The location could not be parsed, show a 404 instead.
      console.error(e);
      dispatch(show404(location.pathname + location.search));
      return;
    }

    // Validate the initial URL state. We can't refresh on a from-file URL.
    if (urlState.dataSource === 'from-file') {
      urlState = null;
    }

    // Normally having multiple dispatches is an anti pattern, but here it's
    // necessary because we are doing different things inside those actions and
    // they can't be merged because we are also calling those seperately on
    // other parts of the code.
    // The first dispatch here updates the url state, then changes state as the url
    // setup is done, and lastly finalizes the profile view since everything is set up now.
    dispatch(updateUrlState(urlState));
    dispatch(urlSetupDone());
    dispatch(finalizeProfileView());
  };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export function reportTrackThreadHeight(
  threadIndex: ThreadIndex,
  height: CssPixels
): ThunkAction<void> {
  return (dispatch, getState) => {
    const trackThreadHeights = getTrackThreadHeights(getState());
    const previousHeight = trackThreadHeights[threadIndex];
    if (previousHeight !== height) {
      // Guard against unnecessary dispatches. This could happen frequently.
      dispatch({
        type: 'UPDATE_TRACK_THREAD_HEIGHT',
        height,
        threadIndex,
      });
    }
  };
}

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}

/**
 * Called when a user has started dragging a file. Used for loading
 * profiles with the drag and drop component.
 */
export function startDragging(): Action {
  return { type: 'START_DRAGGING' };
}

/**
 * Called when a user has stopped dragging a file.
 */
export function stopDragging(): Action {
  return { type: 'STOP_DRAGGING' };
}

/**
 * Called when a custom drag and drop overlay is mounted. This lets
 * the app know that we shouldn't create a default overlay.
 */
export function registerDragAndDropOverlay(): Action {
  return { type: 'REGISTER_DRAG_AND_DROP_OVERLAY' };
}

/**
 * Called when a custom drag and drop overlay is unmounted.
 */
export function unregisterDragAndDropOverlay(): Action {
  return { type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY' };
}
