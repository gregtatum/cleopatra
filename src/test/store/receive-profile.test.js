/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import sinon from 'sinon';
import { blankStore } from '../fixtures/stores';
import * as ProfileViewSelectors from '../../reducers/profile-view';
import * as UrlStateSelectors from '../../reducers/url-state';
import { getView } from '../../reducers/app';
import {
  receiveProfileFromAddon,
  retrieveProfileFromAddon,
  retrieveProfileFromStore,
  retrieveProfileFromUrl,
  _fetchProfile,
} from '../../actions/receive-profile';

import preprocessedProfile from '../fixtures/profiles/profile-2d-canvas.json';
import getGeckoProfile from '../fixtures/profiles/gecko-profile';
import { getEmptyProfile } from '../../profile-logic/profile-data';
import JSZip from 'jszip';
import { serializeProfile } from '../../profile-logic/process-profile';
import type { InitJSZip } from 'jszip';

describe('actions/receive-profile', function() {
  /**
   * This function allows to observe all state changes in a Redux store while
   * something's going on.
   * @param {ReduxStore} store
   * @param {() => Promise<any>} func Process that will be started while
   * observing the store.
   * @returns {Promise<State[]>} All states that happened while waiting for
   * the end of func.
   */
  async function observeStoreStateChanges(store, func) {
    const states = [];
    const unsubscribe = store.subscribe(() => {
      states.push(store.getState());
    });

    await func();

    unsubscribe();
    return states;
  }

  describe('receiveProfileFromAddon', function() {
    it('can take a profile from an addon and save it to state', function() {
      const store = blankStore();

      const initialProfile = ProfileViewSelectors.getProfile(store.getState());
      expect(initialProfile).toBeTruthy();
      expect(initialProfile.threads).toHaveLength(0);
      store.dispatch(receiveProfileFromAddon(preprocessedProfile));
      expect(ProfileViewSelectors.getProfile(store.getState())).toBe(
        preprocessedProfile
      );
    });
  });

  describe('retrieveProfileFromAddon', function() {
    let geckoProfiler;
    let clock;

    beforeEach(function() {
      clock = sinon.useFakeTimers();

      geckoProfiler = {
        getProfile: () => Promise.resolve(getGeckoProfile()),
        getSymbolTable: () => Promise.resolve(),
      };
      window.geckoProfilerPromise = Promise.resolve(geckoProfiler);
    });

    afterEach(function() {
      clock.restore();

      geckoProfiler = null;
      delete window.geckoProfilerPromise;
    });

    it('can retrieve a profile from the addon', async function() {
      const store = blankStore();
      await store.dispatch(retrieveProfileFromAddon());

      const state = store.getState();
      expect(getView(state)).toEqual({ phase: 'PROFILE' });
      expect(ProfileViewSelectors.getDisplayRange(state)).toEqual({
        start: 0,
        end: 1007,
      });
      expect(UrlStateSelectors.getThreadOrder(state)).toEqual([0, 2, 1]); // 1 is last because it's the Compositor thread
      expect(ProfileViewSelectors.getProfile(state).threads).toHaveLength(3); // not empty
    });

    it('displays a warning after 30 seconds', async function() {
      const store = blankStore();

      const states = await observeStoreStateChanges(store, () => {
        const dispatchResultPromise = store.dispatch(
          retrieveProfileFromAddon()
        );
        clock.tick(30000); // this will triggers the timeout synchronously, before the profiler promise's then is run.
        return dispatchResultPromise;
      });
      const views = states.map(state => getView(state));

      const errorMessage =
        'We were unable to connect to the Gecko profiler add-on within thirty seconds. This might be because the profile is big or your machine is slower than usual. Still waiting...';

      expect(views.slice(0, 3)).toEqual([
        {
          phase: 'INITIALIZING',
          additionalData: { attempt: null, message: errorMessage },
        }, // when the error happens
        { phase: 'INITIALIZING' }, // when we could connect to the addon but waiting for the profile
        { phase: 'PROFILE' }, // yay, we got a profile!
      ]);

      const state = store.getState();
      expect(getView(state)).toEqual({ phase: 'PROFILE' });
      expect(ProfileViewSelectors.getDisplayRange(state)).toEqual({
        start: 0,
        end: 1007,
      });
      expect(UrlStateSelectors.getThreadOrder(state)).toEqual([0, 2, 1]); // 1 is last because it's the Compositor thread
      expect(ProfileViewSelectors.getProfile(state).threads).toHaveLength(3); // not empty
    });
  });

  describe('retrieveProfileFromStore', function() {
    const fetch403Response = { ok: false, status: 403 };
    const fetch500Response = { ok: false, status: 500 };
    const fetch200Response = {
      ok: true,
      status: 200,
      headers: {
        get: () => 'appliciation/json',
      },
      json: () => Promise.resolve(getGeckoProfile()),
    };

    beforeEach(function() {
      // The stub makes it easy to return different values for different
      // arguments. Here we define the default return value because there is no
      // argument specified.
      window.fetch = sinon.stub();
      window.fetch.resolves(fetch403Response);

      sinon.stub(window, 'setTimeout').yieldsAsync(); // will call its argument asynchronously
    });

    afterEach(function() {
      delete window.fetch;
      window.setTimeout.restore();
    });

    it('can retrieve a profile from the web and save it to state', async function() {
      const hash = 'c5e53f9ab6aecef926d4be68c84f2de550e2ac2f';
      const expectedUrl = `https://profile-store.commondatastorage.googleapis.com/${hash}`;
      window.fetch.withArgs(expectedUrl).resolves(fetch200Response);

      const store = blankStore();
      await store.dispatch(retrieveProfileFromStore(hash));

      const state = store.getState();
      expect(getView(state)).toEqual({ phase: 'PROFILE' });
      expect(ProfileViewSelectors.getDisplayRange(state)).toEqual({
        start: 0,
        end: 1007,
      });
      expect(UrlStateSelectors.getThreadOrder(state)).toEqual([0, 2, 1]); // 1 is last because it's the Compositor thread
      expect(ProfileViewSelectors.getProfile(state).threads.length).toBe(3); // not empty
    });

    it('requests several times in case of 403', async function() {
      const hash = 'c5e53f9ab6aecef926d4be68c84f2de550e2ac2f';
      const expectedUrl = `https://profile-store.commondatastorage.googleapis.com/${hash}`;
      // The first call will still be a 403 -- remember, it's the default return value.
      window.fetch
        .withArgs(expectedUrl)
        .onSecondCall()
        .resolves(fetch200Response);

      const store = blankStore();
      const views = (await observeStoreStateChanges(store, () =>
        store.dispatch(retrieveProfileFromStore(hash))
      )).map(state => getView(state));

      const errorMessage = 'Profile not found on remote server.';
      expect(views).toEqual([
        { phase: 'INITIALIZING' },
        {
          phase: 'INITIALIZING',
          additionalData: {
            attempt: { count: 1, total: 11 },
            message: errorMessage,
          },
        },
        { phase: 'PROFILE' },
      ]);

      const state = store.getState();
      expect(ProfileViewSelectors.getDisplayRange(state)).toEqual({
        start: 0,
        end: 1007,
      });
      expect(UrlStateSelectors.getThreadOrder(state)).toEqual([0, 2, 1]); // 1 is last because it's the Compositor thread
      expect(ProfileViewSelectors.getProfile(state).threads.length).toBe(3); // not empty
    });

    it('fails in case the profile cannot be found after several tries', async function() {
      const hash = 'c5e53f9ab6aecef926d4be68c84f2de550e2ac2f';
      const store = blankStore();
      const views = (await observeStoreStateChanges(store, () =>
        store.dispatch(retrieveProfileFromStore(hash))
      )).map(state => getView(state));

      const steps = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

      const errorMessage = 'Profile not found on remote server.';
      expect(views).toEqual([
        { phase: 'INITIALIZING' },
        ...steps.map(step => ({
          phase: 'INITIALIZING',
          additionalData: {
            attempt: { count: step, total: 11 },
            message: errorMessage,
          },
        })),
        { phase: 'FATAL_ERROR', error: expect.any(Error) },
      ]);
    });

    it('fails in case the fetch returns a server error', async function() {
      const hash = 'c5e53f9ab6aecef926d4be68c84f2de550e2ac2f';
      window.fetch.resolves(fetch500Response);

      const store = blankStore();
      await store.dispatch(retrieveProfileFromStore(hash));
      expect(getView(store.getState())).toEqual({
        phase: 'FATAL_ERROR',
        error: expect.any(Error),
      });
    });
  });

  describe('retrieveProfileFromUrl', function() {
    const fetch403Response = { ok: false, status: 403 };
    const fetch500Response = { ok: false, status: 500 };
    const fetch200Response = {
      ok: true,
      status: 200,
      headers: {
        get: () => 'application/json',
      },
      json: () => Promise.resolve(getGeckoProfile()),
    };

    beforeEach(function() {
      // The stub makes it easy to return different values for different
      // arguments. Here we define the default return value because there is no
      // argument specified.
      window.fetch = sinon.stub();
      window.fetch.resolves(fetch403Response);

      sinon.stub(window, 'setTimeout').yieldsAsync(); // will call its argument asynchronously
    });

    afterEach(function() {
      delete window.fetch;
      window.setTimeout.restore();
    });

    it('can retrieve a profile from the web and save it to state', async function() {
      const expectedUrl = 'https://profiles.club/shared.json';
      window.fetch.withArgs(expectedUrl).resolves(fetch200Response);

      const store = blankStore();
      await store.dispatch(retrieveProfileFromUrl(expectedUrl));

      const state = store.getState();
      expect(getView(state)).toEqual({ phase: 'PROFILE' });
      expect(ProfileViewSelectors.getDisplayRange(state)).toEqual({
        start: 0,
        end: 1007,
      });
      expect(UrlStateSelectors.getThreadOrder(state)).toEqual([0, 2, 1]); // 1 is last because it's the Compositor thread
      expect(ProfileViewSelectors.getProfile(state).threads.length).toBe(3); // not empty
    });

    it('requests several times in case of 403', async function() {
      const expectedUrl = 'https://profiles.club/shared.json';
      // The first call will still be a 403 -- remember, it's the default return value.
      window.fetch
        .withArgs(expectedUrl)
        .onSecondCall()
        .resolves(fetch200Response);

      const store = blankStore();
      const views = (await observeStoreStateChanges(store, () =>
        store.dispatch(retrieveProfileFromUrl(expectedUrl))
      )).map(state => getView(state));

      const errorMessage = 'Profile not found on remote server.';
      expect(views).toEqual([
        { phase: 'INITIALIZING' },
        {
          phase: 'INITIALIZING',
          additionalData: {
            attempt: { count: 1, total: 11 },
            message: errorMessage,
          },
        },
        { phase: 'PROFILE' },
      ]);

      const state = store.getState();
      expect(ProfileViewSelectors.getDisplayRange(state)).toEqual({
        start: 0,
        end: 1007,
      });
      expect(UrlStateSelectors.getThreadOrder(state)).toEqual([0, 2, 1]); // 1 is last because it's the Compositor thread
      expect(ProfileViewSelectors.getProfile(state).threads.length).toBe(3); // not empty
    });

    it('fails in case the profile cannot be found after several tries', async function() {
      const expectedUrl = 'https://profiles.club/shared.json';
      const store = blankStore();
      const views = (await observeStoreStateChanges(store, () =>
        store.dispatch(retrieveProfileFromUrl(expectedUrl))
      )).map(state => getView(state));

      const steps = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

      const errorMessage = 'Profile not found on remote server.';
      expect(views).toEqual([
        { phase: 'INITIALIZING' },
        ...steps.map(step => ({
          phase: 'INITIALIZING',
          additionalData: {
            attempt: { count: step, total: 11 },
            message: errorMessage,
          },
        })),
        { phase: 'FATAL_ERROR', error: expect.any(Error) },
      ]);
    });

    it('fails in case the fetch returns a server error', async function() {
      const expectedUrl = 'https://profiles.club/shared.json';
      window.fetch.resolves(fetch500Response);

      const store = blankStore();
      await store.dispatch(retrieveProfileFromUrl(expectedUrl));
      expect(getView(store.getState())).toEqual({
        phase: 'FATAL_ERROR',
        error: expect.any(Error),
      });
    });
  });

  /**
   * _fetchProfile is a helper function for the actions, but it is tested separately
   * since it has a decent amount of complexity around different issues with loading
   * in different support URL formats. It's mainly testing what happens when JSON
   * and zip file is sent, and what happens when things fail.
   */
  describe('_fetchProfile', function() {
    beforeEach(function() {
      window.fetch = sinon.stub();
      sinon.stub(window, 'setTimeout').yieldsAsync(); // will call its argument asynchronously
    });

    afterEach(function() {
      delete window.fetch;
      window.setTimeout.restore();
    });

    /**
     * This helper function encapsulates various configurations for the type of content
     * as well and response headers.
     */
    async function configureFetch(obj: {
      url: string,
      contentType?: string,
      isZipped?: true,
      isJSON?: true,
      arrayBuffer?: () => Promise<Uint8Array>,
      json?: () => Promise<mixed>,
    }) {
      const { url, contentType, isZipped, isJSON } = obj;
      const profile = getEmptyProfile();
      let arrayBuffer = obj.arrayBuffer;
      let json = obj.json;

      if (isZipped) {
        const zip = new (JSZip: InitJSZip)();
        zip.file('profile.json', serializeProfile(profile));
        const buffer = await zip.generateAsync({ type: 'uint8array' });
        arrayBuffer = () => buffer;
        json = () => Promise.reject(new Error('Not JSON'));
      }

      if (isJSON) {
        arrayBuffer = () => Promise.reject(new Error('Unhandled mock'));
        json = () => Promise.resolve(profile);
      }

      const zippedProfileResponse = {
        ok: true,
        status: 200,
        json,
        arrayBuffer,
        headers: {
          get: (name: string) => {
            switch (name) {
              case 'content-type':
                return contentType;
              default:
                throw new Error(
                  "Unhandled stub for fetch's response.headers.get"
                );
            }
          },
        },
      };
      window.fetch.withArgs(url).resolves(zippedProfileResponse);
      const reportError = jest.fn();
      const args = {
        url,
        onTemporaryError: () => {},
        reportError,
      };

      // Return fetch's args, based on the inputs.
      return { profile, args, reportError };
    }

    it('fetches a normal profile with the correct content-type headers', async function() {
      const { profile, args } = await configureFetch({
        url: 'https://example.com/profile.json',
        contentType: 'application/json',
        isJSON: true,
      });

      const profileFetched = await _fetchProfile(args);
      expect(profileFetched).toEqual(profile);
    });

    it('fetches a zipped profile with correct content-type headers', async function() {
      const { profile, args, reportError } = await configureFetch({
        url: 'https://example.com/profile.zip',
        contentType: 'application/zip',
        isZipped: true,
      });

      const profileFetched = await _fetchProfile(args);
      expect(profileFetched).toEqual(profile);
      expect(reportError.mock.calls.length).toBe(0);
    });

    it('fetches a zipped profile with incorrect content-type headers, but .zip extension', async function() {
      const { profile, args, reportError } = await configureFetch({
        url: 'https://example.com/profile.zip',
        isZipped: true,
      });

      const profileFetched = await _fetchProfile(args);
      expect(profileFetched).toEqual(profile);
      expect(reportError.mock.calls.length).toBe(0);
    });

    it('fetches a profile with incorrect content-type headers, but .json extension', async function() {
      const { profile, args, reportError } = await configureFetch({
        url: 'https://example.com/profile.json',
        isJSON: true,
      });

      const profileFetched = await _fetchProfile(args);
      expect(profileFetched).toEqual(profile);
      expect(reportError.mock.calls.length).toBe(0);
    });

    it('fetches a profile with incorrect content-type headers, no known extension, and attempts to JSON parse it it', async function() {
      const { profile, args, reportError } = await configureFetch({
        url: 'https://example.com/profile.file',
        isJSON: true,
      });

      const profileFetched = await _fetchProfile(args);
      expect(profileFetched).toEqual(profile);
      expect(reportError.mock.calls.length).toBe(0);
    });

    it('fails if a bad zip file is passed in', async function() {
      const { args, reportError } = await configureFetch({
        url: 'https://example.com/profile.file',
        contentType: 'application/zip',
        arrayBuffer: () => Promise.resolve(new Uint8Array([0, 1, 2, 3])),
      });

      let userFacingError;
      try {
        await _fetchProfile(args);
      } catch (error) {
        userFacingError = error;
      }
      expect(userFacingError).toMatchSnapshot();
      expect(reportError.mock.calls.length).toBeGreaterThan(0);
      expect(reportError.mock.calls).toMatchSnapshot();
    });

    it('fails if a bad profile JSON is passed in', async function() {
      const { args, reportError } = await configureFetch({
        url: 'https://example.com/profile.json',
        contentType: 'application/json',
        json: () => Promise.reject(new Error('Invalid JSON')),
      });

      let userFacingError;
      try {
        await _fetchProfile(args);
      } catch (error) {
        userFacingError = error;
      }
      expect(userFacingError).toMatchSnapshot();
      expect(reportError.mock.calls.length).toBeGreaterThan(0);
      expect(reportError.mock.calls).toMatchSnapshot();
    });

    it('fails if a bad profile JSON is passed in, with no content type', async function() {
      const { args, reportError } = await configureFetch({
        url: 'https://example.com/profile.json',
        json: () => Promise.reject(new Error('Invalid JSON')),
      });

      let userFacingError;
      try {
        await _fetchProfile(args);
      } catch (error) {
        userFacingError = error;
      }
      expect(userFacingError).toMatchSnapshot();
      expect(reportError.mock.calls.length).toBeGreaterThan(0);
      expect(reportError.mock.calls).toMatchSnapshot();
    });

    it('fails if a completely unknown file is passed in', async function() {
      const { args, reportError } = await configureFetch({
        url: 'https://example.com/profile.unknown',
        json: () => Promise.reject(new Error('Invalid JSON')),
      });

      let userFacingError;
      try {
        await _fetchProfile(args);
      } catch (error) {
        userFacingError = error;
      }
      expect(userFacingError).toMatchSnapshot();
      expect(reportError.mock.calls.length).toBeGreaterThan(0);
      expect(reportError.mock.calls).toMatchSnapshot();
    });
  });
});
