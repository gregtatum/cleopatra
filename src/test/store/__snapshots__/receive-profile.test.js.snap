// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions/receive-profile _fetchProfile fails if a bad profile JSON is passed in 1`] = `[Error: The profile’s JSON could not be decoded. The full error information has been printed out to the DevTool’s console.]`;

exports[`actions/receive-profile _fetchProfile fails if a bad profile JSON is passed in 2`] = `
Array [
  Array [
    "The profile’s JSON could not be decoded.",
  ],
  Array [
    "JSON parsing error:",
    [Error: Invalid JSON],
  ],
  Array [
    "Fetch response:",
    Object {
      "arrayBuffer": undefined,
      "headers": Object {
        "get": [Function],
      },
      "json": [Function],
      "ok": true,
      "status": 200,
    },
  ],
]
`;

exports[`actions/receive-profile _fetchProfile fails if a bad profile JSON is passed in, with no content type 1`] = `[Error: The profile’s JSON could not be decoded. The file was not sent with the "application/json" content type, but it did have a .json file ending. Are you sure it was actually JSON? The full error information has been printed out to the DevTool’s console.]`;

exports[`actions/receive-profile _fetchProfile fails if a bad profile JSON is passed in, with no content type 2`] = `
Array [
  Array [
    "The profile’s JSON could not be decoded. The file was not sent with the \\"application/json\\" content type, but it did have a .json file ending. Are you sure it was actually JSON?",
  ],
  Array [
    "JSON parsing error:",
    [Error: Invalid JSON],
  ],
  Array [
    "Fetch response:",
    Object {
      "arrayBuffer": undefined,
      "headers": Object {
        "get": [Function],
      },
      "json": [Function],
      "ok": true,
      "status": 200,
    },
  ],
]
`;

exports[`actions/receive-profile _fetchProfile fails if a bad zip file is passed in 1`] = `[Error: Unable to unzip the zip file. The full error information has been printed out to the DevTool’s console.]`;

exports[`actions/receive-profile _fetchProfile fails if a bad zip file is passed in 2`] = `
Array [
  Array [
    "Unable to unzip the zip file.",
  ],
  Array [
    "Error:",
    [Error: Can't find end of central directory : is this a zip file ? If it is, see https://stuk.github.io/jszip/documentation/howto/read_zip.html],
  ],
  Array [
    "Fetch response:",
    Object {
      "arrayBuffer": [Function],
      "headers": Object {
        "get": [Function],
      },
      "json": undefined,
      "ok": true,
      "status": 200,
    },
  ],
]
`;

exports[`actions/receive-profile _fetchProfile fails if a completely unknown file is passed in 1`] = `[Error: The profile could not be decoded. This does not look like a supported file type. The full error information has been printed out to the DevTool’s console.]`;

exports[`actions/receive-profile _fetchProfile fails if a completely unknown file is passed in 2`] = `
Array [
  Array [
    "The profile could not be decoded. This does not look like a supported file type.",
  ],
  Array [
    "JSON parsing error:",
    [Error: Invalid JSON],
  ],
  Array [
    "Fetch response:",
    Object {
      "arrayBuffer": undefined,
      "headers": Object {
        "get": [Function],
      },
      "json": [Function],
      "ok": true,
      "status": 200,
    },
  ],
]
`;
